spring-commercetools/src/main/java/com/epam/esm/commercetools/certificate/UpdateActionsBuilder.java

```diff
@@ -0,0 +1,119 @@
+package com.epam.esm.commercetools.certificate;
+
+import com.commercetools.api.models.common.LocalizedString;
+import com.commercetools.api.models.common.Money;
+import com.commercetools.api.models.common.PriceDraft;
+import com.commercetools.api.models.product.Attribute;
+import com.commercetools.api.models.product.ProductUpdateAction;
+import com.epam.esm.giftcertficate.model.GiftCertificate;
+import com.epam.esm.tag.model.Tag;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Component
+public class UpdateActionsBuilder {
+
+    private static final String CURRENCY_CODE = "UAH";
+    private static final String DURATION = "Duration";
+    private static final String TAGS = "Tags";
+    private final Map<Predicate<GiftCertificate>, Consumer<List<ProductUpdateAction>>> updatesMap = new HashMap<>();
+    private Attribute setDurationAttribute(int duration) {
+        return Attribute
+                .builder()
+                .name(DURATION)
+                .value(duration)
+                .build();
+
+    }
+
+    private Attribute setTagsAttributeWhenUpdateGiftCertificate(Set<Tag> tags) {
+        if (tags.isEmpty()) {
+            return Attribute
+                    .builder()
+                    .name(TAGS)
+                    .value(List.of())
+                    .build();
+        }
+
+        return Attribute
+                .builder()
+                .name(TAGS)
+                .value(
+                        tags.stream().map(Tag::getName)
+                                .collect(Collectors.toSet())
+                )
+                .build();
+    }
+
+    public List<ProductUpdateAction> preparingUpdatesForGiftCertificate(GiftCertificate giftCertificateUpdates) {
+        List<ProductUpdateAction> productUpdateActions = new ArrayList<>();
+
+        updatesMap.put(giftCertificate -> giftCertificateUpdates.getName() != null,
```

Comment:
Those `put` calls shouldn't be in this method too. You are re-populating map for each update, but need to do it once, at the application startup and then just use the map, as you do in loop on line 109

Reply:
i use giftcertificate , how can i remake this ?

Reply:
I have already remade this one